crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -2001,10 +2037,69 @@ mod tests {
     }
 
     #[test]
-    fn test_block_num_reader() -> eyre::Result<()> {
+    fn test_withdrawals_provider() -> eyre::Result<()> {
         let mut rng = generators::rng();
+        let chain_spec = Arc::new(ChainSpecBuilder::mainnet().shanghai_activated().build());
         let (provider, database_blocks, in_memory_blocks, _) =
-            provider_with_random_blocks(&mut rng, TEST_BLOCKS_COUNT, TEST_BLOCKS_COUNT, None)?;
+            provider_with_chain_spec_and_random_blocks(
+                &mut rng,
+                chain_spec.clone(),
+                TEST_BLOCKS_COUNT,
+                TEST_BLOCKS_COUNT,
+                None,
+                Some(1..3),
+            )?;
+        let blocks = [database_blocks, in_memory_blocks].concat();
+
+        let shainghai_timestamp =
+            chain_spec.hardforks.fork(EthereumHardfork::Shanghai).as_timestamp().unwrap();
+
+        assert_eq!(
+            provider
+                .withdrawals_by_block(
+                    reth_primitives::BlockHashOrNumber::Number(15),
+                    shainghai_timestamp
+                )
+                .expect("could not call withdrawals by block"),
+            Some(Withdrawals::new(vec![])),
+            "Expected withdrawals_by_block to return empty list if block does not exist"
+        );
+
+        for i in 0..TEST_BLOCKS_COUNT {
```

Comment:
I think it only queries database blocks, because we insert `TEST_BLOCKS_COUNT` database + `TEST_BLOCKS_COUNT` in-memory blocks in `provider_with_chain_spec_and_random_blocks`. Maybe we can just do
```suggestion
        for block in blocks {
```
and then get the block number from them to query the provider?

Reply:
thanks for the catch, fixed!

Reply:
not needed , removed it